### ISSUE TITLE
Session.run() on Operation return None. Design question

### ISSUE THREAD

**Issue by Jylomaki** [tensorflow/tensorflow#21053]
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
Yes, see this issue. https://github.com/tensorflow/tensorflow/issues/20990

The documentation of Session.run from tensorflow:
```
run(
    fetches,
    feed_dict=None,
    options=None,
    run_metadata=None
)
```
> Runs operations and evaluates tensors in fetches.
> 
> This method runs one "step" of TensorFlow computation, by running the necessary graph fragment to execute every Operation and evaluate every Tensor in fetches, substituting the values in feed_dict for the corresponding input values.
> 
> The fetches argument may be a single graph element, or an arbitrarily nested list, tuple, namedtuple, dict, or OrderedDict containing graph elements at its leaves. A graph element can be one of the following types:
> 
>     **An tf.Operation. The corresponding fetched value will be None.**
>     A tf.Tensor. The corresponding fetched value will be a numpy ndarray containing the value of that tensor.
>     A tf.SparseTensor. The corresponding fetched value will be a tf.SparseTensorValue containing the value of that sparse tensor.
>     A get_tensor_handle op. The corresponding fetched value will be a numpy ndarray containing the handle of that tensor.
>     A string which is the name of a tensor or operation in the graph.

I was stuck for quite a while on this issue, due to a` sess.run() `returning `None` all the time and not raising any exception.
I don't quite get the design intention under such a behaviour:
`Session.run()` is expected to return values. But it return `None` for `tf.operation`
Not raising any exception, nor outputing any information make you think it is okay to call `sess.run()` on your parameter and that the issue is not on this function call.
Why is it okay to allow user to call this method with parameter that will result in no effect?
Also, as `tf.Operation` is a tensorflow type, wouldn't it be possible to automatically call for its `.values()` in the `Session.run()` instead of returning a `None`?

**Comment by ppwwyyxx** [comment]
> Why is it okay to allow user to call this method with parameter that will result in no effect?

Returning `None` does not mean resulting in no effect. Calling an op has an effect.

> wouldn't it be possible to automatically call for its .values() in the Session.run() instead of returning a None?

It won't be possible because not every op has output values.

**Comment by tanzhenyu** [comment]
There are many types of tf.Operation, some of them does return None, and it serves as grouping ops together, example being training ops returned by optimizer.minimize. Just returning nothing does not mean it's doing nothing, it actually run all the ops that are being grouped.

**Comment by Jylomaki** [comment]
> I'm not sure what ops silently do nothing and return nothing, do you have an example?

Doesn't running a `predict node`  via `node` instead of `node.values` does that?
Maybe it computes the value internally, but from the user standpoint: the `state of the model` didn't change and `no value was returned` (technically `None` was returned).

> There can be `while_loops` with zero iterations, for instance, but sometimes they are there for a reason

I guess they can do nothing, but they have a **possibility** of doing something. (or if they don't, maybe an a hint at the uselessness of the node would be nice)

> There are many types of` tf.Operation`, some of them does return `None`, and it serves as `grouping ops `together, example being `training ops` returned by `optimizer.minimize`. Just returning nothing does not mean it's doing nothing, it actually `run` all the `ops` that are being grouped.

You are right about it **currently** returning `None`. 
But isn't it be plausible to return a `status` of the node operation? like the type of node or such?
I get that it may be desirable for its added knowledge of state of execution but not so much for its overhead cost.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 89 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 105 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 120 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tilakrayal** [comment]
@Jylomaki,
Is this still an issue?
Could you update TensorFlow to the latest stable version v.2.9 and let us know if you are facing the same error. Thanks!
